components/backup/src/disk_snap.rs

```diff
@@ -0,0 +1,340 @@
+            rejector,
+            active_stream: Arc::new(AtomicU64::new(0)),
+            runtime: Either::Right(Self::default_runtime()),
+        }
+    }
+
+    pub fn with_rejector_and_runtime(
+        handle: SR,
+        rejector: Arc<RejectIngestAndAdmin>,
+        runtime: Handle,
+    ) -> Self {
+        Self {
+            handle,
+            rejector,
+            active_stream: Arc::new(AtomicU64::new(0)),
+            runtime: Either::Left(runtime),
+        }
+    }
+
+    fn check_initialized(&self) -> Result<()> {
+        if !self.rejector.connected() {
+            return Err(Error::Uninitialized);
+        }
+        Ok(())
+    }
+
+    fn check_rejected(&self) -> Result<()> {
+        self.check_initialized()?;
+        if self.rejector.allowed() {
+            return Err(Error::LeaseExpired);
+        }
+        Ok(())
+    }
+
+    fn update_lease(&self, lease_dur: Duration) -> Result<PResp> {
+        self.check_initialized()?;
+        let mut event = PResp::new();
+        event.set_ty(PEvnT::UpdateLeaseResult);
+        event.set_last_lease_is_valid(self.rejector.heartbeat(lease_dur));
+        Ok(event)
+    }
+
+    fn reset(&self) -> PResp {
+        let rejected = !self.rejector.allowed();
+        self.rejector.reset();
+        let mut event = PResp::new();
+        event.set_ty(PEvnT::UpdateLeaseResult);
+        event.set_last_lease_is_valid(rejected);
+        event
+    }
+}
+
+pub struct StreamHandleLoop<SR: SnapshotBrHandle + 'static> {
+    pending_regions: Vec<BoxFuture<'static, (Region, Result<()>)>>,
+    env: Env<SR>,
+}
+
+impl<SR: SnapshotBrHandle + 'static> Drop for StreamHandleLoop<SR> {
+    fn drop(&mut self) {
+        self.env.active_stream.fetch_sub(1, Ordering::SeqCst);
+    }
+}
+
+enum StreamHandleEvent {
+    Req(PReq),
+    WaitApplyDone(Region, Result<()>),
+    ConnectionGone(Option<grpcio::Error>),
+}
+
+impl<SR: SnapshotBrHandle + 'static> StreamHandleLoop<SR> {
+    pub fn new(env: Env<SR>) -> Self {
+        env.active_stream.fetch_add(1, Ordering::SeqCst);
+        Self {
+            env,
+            pending_regions: vec![],
+        }
+    }
+
+    fn async_wait_apply(&mut self, region: &Region) -> BoxFuture<'static, (Region, Result<()>)> {
+        if let Err(err) = self.env.check_rejected() {
+            return Box::pin(future::ready((region.clone(), Err(err))));
+        }
+
+        let (tx, rx) = oneshot::channel();
+        let syncer = SnapshotBrWaitApplySyncer::new(region.id, tx);
+        let handle = self.env.handle.clone();
+        let region = region.clone();
+        let epoch = region.get_region_epoch().clone();
+        let id = region.get_id();
+        let send_res = handle
+            .send_wait_apply(id, SnapshotBrWaitApplyRequest::strict(syncer, epoch))
+            .map_err(Error::RaftStore);
+        Box::pin(
+            async move {
+                send_res?;
+                rx.await.map_err(|_| Error::WaitApplyAborted).map(|_| ())
```

Comment:
Why not remove the reigon id from `pending_peers` here? Then you don't need to write `wait_applies` logic in `next_event`

Reply:
I think concurrency access to the same variable may make things more complex. Also I think it would be harder to be notified once some wait apply finished (perhaps we have to poll the status of `pending_peers`). 🤔